### PR TITLE
New transaction locking mechanisms

### PR DESCRIPTION
**Pull request by kshlm**
This PR introduces a new NewTxnWithLocks function and a (*Txn).Lock method.

With this change, support for having locking as part of transaction steps has
been removed. Doing locks within steps could lead to TOCTOU race problems, as
in several cases, volinfos would be fetched/modified in the rest handler before
the transaction steps were performed.

@prashanthpai started doing changes with regards to the TOCTOU problem with
them.

With this change, users now need to explicitly create a transaction and obtain
locks on the required IDs before fetching/modifying stored data in the rest
handlers.

The rest of the transaction framework remains as is.

### CODE REVIEW

File: glusterd2/transaction/transaction.go [hunk 1]
```diff
@@ -39,19 +42,43 @@ type Txn struct {
...
+	}
+
+	return t, nil
+}
+
+// Done releases any obtained locks and cleans up the transaction namespace
+// Done must be called after a transaction ends
+func (t *Txn) Done() {
+	// Release obtained locks
+	for _, locker := range t.locks {
```
Comment (prashanthpai) [inline glusterd2/transaction/transaction.go]: Does the ordering in list of locks matter ? I was wondering whether the last lock taken should be the first to get unlocked.

Comment (kshlm) [reply glusterd2/transaction/transaction.go]: Not really ATM, and I don't expect it to. The locks are taken for the lifetime of the period of the transaction, and are released all at once at the end. I don't expect it to change.

**Comment by kshlm** [comment]
> CreateLockFuncs was a temporary solution to take locks without a transaction i.e there are places where we need locks but no other steps are needed. One such example is editing volume.Metadata or peer.Metadata. Does this PR handle those cases too ? IOW, can you confirm that the new lock functions can be called without populating steps and calling txn.Do() ?

Yes. You can create a new transaction without populating the steps or running the steps. And after that you can add as many locks as you want with txn.Lock(). You need to ensure that you call txn.Done() when you're done.

**Comment by prashanthpai** [comment]
Closes https://github.com/gluster/glusterd2/issues/510

**Comment by prashanthpai** [comment]
retest this please

**Comment by kshlm** [comment]
Okay. This was fast. I was hoping to wait till a few of the other larger PRs got merged (rebalance, snapshot, IVC).

**Pull request merged**